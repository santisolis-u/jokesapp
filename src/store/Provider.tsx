import React, {useEffect, useState} from 'react';
import {Joke} from '../hooks/useJokes';
import {FavJokesContext} from './store';
import Realm from 'realm';
import {JokeSchema} from '../realm/schema';
interface Props {
  children: JSX.Element;
}

export const FavsProvider: React.FC<Props> = ({children}) => {
  const [favJokes, setFavJokes] = useState<Joke[]>([]);
  const [realm, setRealm] = useState<Realm>();

  useEffect(() => {
    const openRealm = async () => {
      const r = await Realm.open({schema: [JokeSchema]});
      if (r != null && !r.isClosed) {
        setRealm(r);
        const result = r.objects<Joke>('Joke').map(value => {
          return value.toJSON() as Joke;
        });
        setFavJokes(result);
      }
    };
    openRealm();
    return () => {
      if (!!realm && !realm.isClosed) {
        realm.close();
      }
    };
  }, []);

  const addFavJoke = (joke: Joke) => {
    if (!favJokes.includes(joke)) {
      if (!!realm && !realm.isClosed) {
        realm.write(() => {
          realm.create('Joke', joke);
        });
      }
      setFavJokes(favJokes => [...favJokes, joke]);
    }
  };
  const removeFavJoke = (id: string) => {
    if (!!realm && !realm.isClosed) {
      realm.write(() => {
        const joke = realm.objectForPrimaryKey<Joke>('Joke', id);
        realm.delete(joke);
      });
    }
    setFavJokes(favJokes => favJokes.filter(joke => joke.id !== id));
  };
  return (
    <FavJokesContext.Provider value={{favJokes, addFavJoke, removeFavJoke}}>
      {children}
    </FavJokesContext.Provider>
  );
};
