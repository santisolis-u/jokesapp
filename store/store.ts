import {Joke} from '../src/hooks/useJokes';
import React, {FC, ReactChildren, useState} from 'react';
interface FavJokesContextType {
  favJokes: Joke[];
  addFavJoke?: (joke: Joke) => void;
  removeFavJoke?: (id: string) => void;
}

const defaultState = {
  favJokes: [],
};

export const FavJokesContext =
  React.createContext<FavJokesContextType>(defaultState);
