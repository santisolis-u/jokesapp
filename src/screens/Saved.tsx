import {FlatList, StyleSheet, Text, View} from 'react-native';
import React, {useContext} from 'react';
import {FavJokesContext} from '../store/store';
import SavedCards from '../components/SavedCards';

const Saved = () => {
  const {favJokes, removeFavJoke} = useContext(FavJokesContext);
  const handleDelete = (id: string) => {
    if (removeFavJoke) {
      removeFavJoke(id);
    }
  };
  return (
    <View style={styles.savedContainer}>
      {favJokes.length > 0 ? (
        <FlatList
          keyExtractor={element => element.id}
          data={favJokes}
          renderItem={joke => (
            <SavedCards handleDelete={handleDelete} joke={joke.item} />
          )}
          style={styles.flatList}
          showsVerticalScrollIndicator={false}
        />
      ) : (
        <View style={styles.center}>
          <Text style={styles.fallbackText}>Nothing saved</Text>
        </View>
      )}
    </View>
  );
};

export default Saved;

const styles = StyleSheet.create({
  savedContainer: {
    margin: 24,
    // overflow: 'visible',
    flex: 1,
  },
  flatList: {
    // overflow: 'visible',
  },
  center: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
  },
  fallbackText: {
    textAlign: 'center',
    fontSize: 32,
    fontFamily: 'OpenSans-Bold',
  },
});
