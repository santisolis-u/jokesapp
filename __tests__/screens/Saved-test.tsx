import React from 'react';
import {MockedProvider} from '@apollo/client/testing';

// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer';
import {fireEvent, render} from '@testing-library/react-native';
import {FavJokesContext} from '../../src/store/store';
import Saved from '../../src/screens/Saved';

const navigation: any = {
  setOptions: jest.fn(),
  navigate: jest.fn(),
};
const mocks: any[] = [];

const removeFavJoke = jest.fn();

describe('Saved', () => {
  test('renders correctly', () => {
    const component = (
      <MockedProvider mocks={mocks}>
        <FavJokesContext.Provider
          value={{
            addFavJoke: jest.fn(),
            removeFavJoke: jest.fn(),
            favJokes: [
              {
                id: '1',
                joke: 'TEST',
              },
            ],
          }}>
          <Saved />
        </FavJokesContext.Provider>
      </MockedProvider>
    );
    const json = renderer.create(component).toJSON();
    expect(json).toMatchSnapshot();
  });
  test('Click remove in card calls context removeJoke', async () => {
    const component = (
      <MockedProvider mocks={mocks}>
        <FavJokesContext.Provider
          value={{
            addFavJoke: jest.fn(),
            removeFavJoke,
            favJokes: [
              {
                id: '1',
                joke: 'Test',
              },
            ],
          }}>
          <Saved />
        </FavJokesContext.Provider>
      </MockedProvider>
    );
    const {getByTestId} = render(component);
    const toClick = await getByTestId('trash');
    fireEvent(toClick, 'press');
    expect(removeFavJoke).toHaveBeenCalled();
  });
});
