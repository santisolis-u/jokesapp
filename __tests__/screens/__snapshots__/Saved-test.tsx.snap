// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Saved renders correctly 1`] = `
<View
  style={
    Object {
      "flex": 1,
      "margin": 24,
      "overflow": "visible",
    }
  }
>
  <RCTScrollView
    data={
      Array [
        Object {
          "id": "1",
          "joke": "TEST",
        },
      ]
    }
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    showsVerticalScrollIndicator={false}
    stickyHeaderIndices={Array []}
    style={
      Object {
        "overflow": "visible",
      }
    }
    viewabilityConfigCallbackPairs={Array []}
  >
    <View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "white",
              "borderColor": "black",
              "borderRadius": 24,
              "borderWidth": 2,
              "flex": 1,
              "flexDirection": "row",
              "fontFamily": "OpenSans-Bold",
              "justifyContent": "center",
              "marginBottom": 16,
              "minHeight": 100,
              "padding": 16,
              "shadowColor": "black",
              "shadowOffset": Object {
                "height": 2,
                "width": 2,
              },
              "shadowOpacity": 0.2,
              "shadowRadius": 24,
            }
          }
        >
          <Text
            style={
              Object {
                "alignItems": "center",
                "flex": 3,
                "fontFamily": "OpenSans-Bold",
                "justifyContent": "center",
              }
            }
          >
            TEST
          </Text>
          <View
            accessible={true}
            collapsable={false}
            focusable={true}
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "alignItems": "center",
                "flex": 1,
                "justifyContent": "center",
              }
            }
          >
            <Text
              style={Object {}}
              testID="trash"
            >
              🗑️
            </Text>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
</View>
`;
